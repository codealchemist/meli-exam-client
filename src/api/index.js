const baseUrl = 'https://api.mercadolibre.com'

export const search = query => {
  console.log('API SEARCH', query)
  return window
    .fetch(`${baseUrl}/sites/MLA/search?q=${query}`)
    .then(res => res.json())
    .catch(err => {
      console.error('Unable search items', err)
    })
}

export const getProductDetails = id => {
  console.log('API get product details', id)
  const getDetails = window
    .fetch(`${baseUrl}/items/${id}`)
    .then(res => res.json())
  const getDescription = window
    .fetch(`${baseUrl}/items/${id}/description`)
    .then(res => res.json())

  return Promise.all([getDetails, getDescription])
    .then(([details, description]) => ({ details, description }))
    .catch(err => {
      console.error('Unable to get product details', err)
    })
}

export const getCategory = id => {
  console.log('API get category', id)
  return window
    .fetch(`${baseUrl}/categories/${id}`)
    .then(res => res.json())
    .catch(err => {
      console.error('Unable to get category', err)
    })
}
