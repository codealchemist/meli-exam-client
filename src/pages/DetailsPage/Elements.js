import styled from 'styled-components'
import Colors from '../../components/Colors'

export const Wrapper = styled.div`
  width: 100%;
  background: ${Colors.default.light2};
  padding-bottom: 40px;
`

export const Content = styled.div`
  width: 100%;
  max-width: 1200px;
  background: ${Colors.default.light2};
  margin: auto;
`
