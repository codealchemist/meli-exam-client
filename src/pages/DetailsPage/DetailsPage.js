import React from 'react'
import styled from 'styled-components'
import Breadcrumb from '../../components/Breadcrumb'
import Context from '../../Context'
import { getCategory as getCategoryApi } from '../../api'
import ProductDetails from '../../components/ProductDetails'
import Colors from '../../components/Colors'
import { getProductDetails as getProductDetailsApi } from '../../api'

const Wrapper = styled.div`
  width: 100%;
  background: ${Colors.default.light2};
  padding-bottom: 40px;
`

const Content = styled.div`
  width: 100%;
  max-width: 1200px;
  background: ${Colors.default.light2};
  margin: auto;
`

class DetailsPage extends React.PureComponent {
  state = {
    searching: false
  }

  componentDidMount() {
    if (!this.context.productDetails) {
      const { params } = this.props.match
      this.getProductDetails(params.id)
      return
    }

    this.getCategory()
  }

  getProductDetails = id => {
    this.setState({ searching: true })
    getProductDetailsApi(id).then(response => {
      console.log('GOT product details response', response)
      this.context.setProductDetails(response)
      this.getCategory()
      this.setState({ searching: false })
    })
  }

  getCategory = () => {
    const { category_id } = this.context.productDetails.details
    getCategoryApi(category_id).then(categories => {
      console.log('got categories', categories)
      this.context.setProductCategories(categories)
    })
  }

  render() {
    return (
      <Context.Consumer>
        {context => {
          if (!context.productDetails) return null

          return (
            <Wrapper>
              <Content>
                <Breadcrumb
                  categories={context.productCategories.path_from_root}
                />
                <ProductDetails product={context.productDetails} />
              </Content>
            </Wrapper>
          )
        }}
      </Context.Consumer>
    )
  }
}

DetailsPage.contextType = Context
export default DetailsPage
