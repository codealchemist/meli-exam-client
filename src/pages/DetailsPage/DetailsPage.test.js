import React from 'react'
import { mount } from 'enzyme'
import DetailsPage from './DetailsPage'
import * as api from '../../api'

// Turn off console log.
// global.console = { log: jest.fn() }

jest.mock('../../Context', () => ({
  Consumer: obj =>
    obj.children({
      productDetails: {},
      productCategories: {
        path_from_root: []
      }
    })
}))
// jest.mock('../../api', () => ({
//   getProductDetails: () => ({
//     then: callback => callback({ id: 42 })
//   }),
//   getCategory: () => ({
//     then: callback => callback({ name: 'awesome-category' })
//   })
// }))
jest.mock('../../components/Breadcrumb', () => 'div')
jest.mock('../../components/ProductDetails', () => 'div')

const context = {
  setProductDetails: jest.fn(),
  setProductCategories: jest.fn(),
  productDetails: {
    item: {
      category_id: 7
    }
  },
  productCategories: {
    path_from_root: []
  }
}
DetailsPage.contextType = React.createContext(context)

describe('DetailsPage', () => {
  it('should render', () => {
    const match = {
      params: {
        id: 99
      }
    }
    const wrapper = mount(<DetailsPage match={match} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should get category', () => {
    jest.spyOn(api, 'getCategory')
    const wrapper = mount(<DetailsPage />)
    expect(api.getCategory).toHaveBeenCalledWith(7)
  })

  xit('should setProductCategories', () => {
    jest.spyOn(api, 'getCategory')
    jest.spyOn(context, 'setProductCategories')
    api.getCategory.mockReturnValue(Promise.resolve([1, 2]))
    const wrapper = mount(<DetailsPage />)
    expect(context.setProductCategories).toHaveBeenCalled()
  })

  it('should get product details', () => {
    const context = {
      setProductDetails: jest.fn(),
      setProductCategories: jest.fn(),
      productDetails: null,
      productCategories: {
        path_from_root: []
      }
    }
    DetailsPage.contextType = React.createContext(context)

    const match = {
      params: {
        id: 99
      }
    }
    jest.spyOn(api, 'getProductDetails')
    api.getProductDetails.mockReturnValue(Promise.resolve({ name: 'Cool!' }))
    const wrapper = mount(<DetailsPage match={match} />)
    expect(api.getProductDetails).toHaveBeenCalledWith(99)
  })

  xit('should set product details', () => {
    const context = {
      setProductDetails: jest.fn(),
      setProductCategories: jest.fn(),
      productDetails: null,
      productCategories: {
        path_from_root: []
      }
    }
    DetailsPage.contextType = React.createContext(context)

    const match = {
      params: {
        id: 99
      }
    }
    jest.spyOn(api, 'getProductDetails')
    jest.spyOn(context, 'setProductDetails')
    api.getProductDetails.mockReturnValue(Promise.resolve({ name: 'Cool!' }))
    const wrapper = mount(<DetailsPage match={match} />)
    expect(context.setProductDetails).toHaveBeenCalledWith()
  })
})
