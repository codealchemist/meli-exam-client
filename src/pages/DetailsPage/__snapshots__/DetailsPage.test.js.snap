// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsPage should render 1`] = `
<DetailsPage
  match={
    Object {
      "params": Object {
        "id": 99,
      },
    }
  }
>
  <Consumer>
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": true,
              "lastClassName": "bDyzBP",
              "rules": Array [
                "
  width: 100%;
  background: ",
                "#eee",
                ";
  padding-bottom: 40px;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-bdVaJa bDyzBP"
        >
          <styled.div>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bwzfXH",
                    "isStatic": true,
                    "lastClassName": "gguNSM",
                    "rules": Array [
                      "
  width: 100%;
  max-width: 1200px;
  background: ",
                      "#eee",
                      ";
  margin: auto;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bwzfXH",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <div
                className="sc-bwzfXH gguNSM"
              >
                <div
                  categories={Array []}
                />
                <div
                  product={Object {}}
                />
              </div>
            </StyledComponent>
          </styled.div>
        </div>
      </StyledComponent>
    </styled.div>
  </Consumer>
</DetailsPage>
`;
